src/redisearch_rs/c_entrypoint/iterators_ffi/src/c2rust.rs

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+// TODO: remove once we have compound iterators written in Rust that leverage
+//   this shim.
+#![allow(unused)]
+use ffi::{
+    IteratorStatus_ITERATOR_EOF, IteratorStatus_ITERATOR_NOTFOUND, IteratorStatus_ITERATOR_OK,
+    IteratorStatus_ITERATOR_TIMEOUT, QueryIterator, t_docId,
+};
+use inverted_index::RSIndexResult;
+use rqe_iterators::{RQEIterator, RQEIteratorError, SkipToOutcome};
+use std::ptr::NonNull;
+
+/// A Rust shim over a query iterator that satisfies the C iterator API.
+///
+/// # Rationale
+///
+/// [`CRQEIterator`] allows Rust code to treat query iterators written in C
+/// as if they implemented the [`RQEIterator`] trait.
+///
+/// This is particularly useful when composing iterators—e.g.
+/// in the union or intersection iterators. They can work seamslessly with both C
+/// and Rust iterators, since they both implement the [`RQEIterator`] trait: Rust
+/// iterators do it "directly", C iterators do it via this shim.
+///
+/// # Implementation details
+///
+/// It is not a given that the underlying iterator is written in C!
+/// It might be a Rust iterator, wrapped to obey the C API, being passed into
+/// a Rust composite iterator.
+#[repr(transparent)]
+pub(crate) struct CRQEIterator {
+    header: QueryIterator,
+}
+
+impl CRQEIterator {
+    /// Transmute a non-null pointer to a C query iterator into a mutable reference
+    /// to its Rust wrapper.
+    ///
+    /// # Safety
+    ///
+    /// 1. It must be safe to cast `NonNull<QueryIterator>` to `&mut QueryIterator`.
+    ///    In particular, the pointer cannot be aliased.
```

Comment:
How is aliasing breaking it?

Reply:
You're casting a raw pointer to a `&mut` reference. If you were to call it twice using the same raw pointer, you'd have two `&mut` references to the same location, thus undefined behaviour.